### ISSUE TITLE
Not enough digit positions for Watermeter counts > 10.000 m3

### ISSUE THREAD

**Issue by bikeymouse** [domoticz/domoticz#2757]
Version: 4.10038 beta (but also latest stable)
Platform: RPI2B Strech 

My watermeter reports per Liter and the total amount used counter is in m3.
So this basically works fine and the counter device (type "RFXMeter counter", type "Water") shows the total counter as "xxx.yyy m3"
That''s nice, because yyy shows the amount in Liters.

However, because my actual watermeter as a total of xxxx.yyy m3 (so more then 10.000 m3), I have added an offset. 
Unfortunately, If I do that the device now shows xxxx.yy m3
So it now rounding up to 10 liters.

I have checked with the JSON API, and this shows that actually Domoticz-core is doing te rounding and that this is not a display/skin-bug.

Issue also reported at the [forum](http://www.domoticz.com/forum/viewtopic.php?f=6&t=25160#p193423)

**Comment by gizmocuz** [comment]
There should be no rounding at all (just checked)

for example when you used 10.880 liters of water, it is displayed as 10.88
if you used 10.881 it will be displayed as 10.881

**Comment by bikeymouse** [comment]
You are right this happens not at 10.000 but at 1000.000 liters (= 1000m3).

So in my case with an offset of 807.511 my usage shows as 807.968m3

However If I change the offset to 1807.511 to reflect my real watermeter, it shows: 1807.97 m3

So it still shows 6 digits effectively losing the least significant digit and rounding up to 10 liters.

**Comment by gizmocuz** [comment]
@bikeymouse , why you closed the issue ? Is it not an issue ?
I use printf "%g" for the value

WebServer.cpp, line 9408
`sprintf(szTmp, "%g m3", musage);`

D you compile yourself ? You can try to change this to %.3f

The same is used for Energy/Gas and i see three digits here

**Comment by bikeymouse** [comment]
Sorry, I unintentionally closed the issue.
Yes it is still an issue. I don’t compile myself. Your suggestion seems to be the right solution (also because even formatting 10.880 liters as 10.88 is also not desirable IMO).

Can this be solved in the base code for everyone?
BTW: I see that my watermeter even has 8 digits.

**Comment by gizmocuz** [comment]
Actually it should not matter how many digits it has, %g should display it, and not round it... it does not round
It is also used for the Gas and Energy devices, do you see devices with 3 decimals ?

**Comment by bikeymouse** [comment]
Good point, I do see 3 decimals for gas and energy meters. The energy meter even has an 8-digit total (of which 3 are decimals), so that works exactly as I like.
If these use the same formatting, then why does that works out differently for the watermeter?

**Comment by gizmocuz** [comment]
unless your watermeter has always a '0' as 3rd decimal like 120 , will be truncated to 12
Could you show me a screenshot of your watermeter widget ?

**Comment by bikeymouse** [comment]
The problem with %g is that if you don’t specify the precision it uses default max 6 significant digits:
“g, G: Double precision value, in %e or %E format if the exponent is less than -4 or greater than or equal to the precision, and %f format otherwise. (The precision (default 6) specifies the number of significant digits here, whereas in %f, %e, it is the number of digits after the decimal point.)”
See: https://stat.ethz.ch/R-manual/R-devel/library/base/html/sprintf.html

**Comment by bikeymouse** [comment]
With offset 807.51:
![abb6a085-488b-4926-9df8-2adf42520ca3](https://user-images.githubusercontent.com/34967010/46756268-574ca880-ccc7-11e8-8f00-01fd5bfb89f5.jpeg)
With offset 1807.51:
![57151fcd-9b86-4c42-9c0c-69ee34d037fd](https://user-images.githubusercontent.com/34967010/46756372-91b64580-ccc7-11e8-80c0-7529065c2c71.jpeg)
And with offset 11807.51 (decimal digits completely lost):
![e37eaebe-3106-4c27-8527-30de29358f72](https://user-images.githubusercontent.com/34967010/46756656-40f31c80-ccc8-11e8-91e1-2e62eace147f.jpeg)

**Comment by gizmocuz** [comment]
Thank you for that information! Did not know this !!! Solved in beta 10056

**Comment by bikeymouse** [comment]
Great, thanks!
Still wondering however why my energy meter does show 8 digits including 3 decimal places, if this uses the same formatting. Can you explain that?
![2560d9ba-3d83-4c53-88ca-ac04c29d8212](https://user-images.githubusercontent.com/34967010/46790635-ae8e5f80-cd3f-11e8-9aad-710e2e5a25d9.jpeg)

**Comment by gizmocuz** [comment]
All formatting has been updated in the last patch. Your widget looks correct, whats wrong with it ?

**Comment by bikeymouse** [comment]
Hi, yes it looks perfect, but that was my point it already did even before the patch! So why was the powermeter working while it used the same code as the watermeter for the formatting.
Anyway, I just updated and can confirm that the watermeter no also looks fine!

**Comment by gizmocuz** [comment]
;) great thanks for the confirmation!

**Comment by gizmocuz** [comment]
This patch caused problems for other sensors... so it's going to be reverted. I will address the water meter differently
https://github.com/domoticz/domoticz/issues/2773#issuecomment-429563003

**Comment by gizmocuz** [comment]
could you try beta 10064, there should be no issue with the watermeter

**Comment by bikeymouse** [comment]
I just tested with beta 4.10067
Everything looks fine on the widgets from the watermeter, flowmeter, gasmeter and energy meter.

Only thing is that the logging of the watermeter does not work. The graph stays empty, and if I select "view data table" I get:
DateTime | Water
2018-10-09 22:00:00 | NaN
2018-10-09 23:00:00 | NaN
2018-10-10 00:00:00 | NaN
2018-10-10 01:00:00 | NaN
etc...

I checked with SQLitebrowser and found that all data in the Meter-table looks Ok to me:
"1009"	"481"	"0"	"2018-10-16 22:05:00"
"1009"	"484"	"0"	"2018-10-16 22:10:00"
"1009"	"497"	"0"	"2018-10-16 22:15:00"
"1009"	"514"	"0"	"2018-10-16 22:20:00"
"1009"	"517"	"0"	"2018-10-16 22:25:00"
"1009"	"527"	"0"	"2018-10-16 22:30:00"
"1009"	"527"	"0"	"2018-10-16 22:35:01"
etc...

Logging from all other meters, even the Waterflowmeter works fine. 
Cleared cache of course and tried in browser private mode and default theme.
